controllers/hyperparametertuningjob/hyperparametertuningjob_controller_test.go

```diff
@@ -18,1680 +18,645 @@ package hyperparametertuningjob
+		sageMakerClient = NewMockSageMakerClientBuilder(GinkgoT()).Build()
 	})
 
-	AfterEach(func() {
-		// Get job so we can delete it.
-		err = k8sClient.Get(context.Background(), types.NamespacedName{
-			Namespace: job.ObjectMeta.Namespace,
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		if err != nil && apierrs.IsNotFound(err) {
-			return
-		}
-		Expect(err).ToNot(HaveOccurred())
-
-		// Assert that deletionTimestamp is nonzero.
-		err = k8sClient.Delete(context.Background(), job)
-		Expect(err).ToNot(HaveOccurred())
-
-		// Get job so we can remove finalizers.
-		err = k8sClient.Get(context.Background(), types.NamespacedName{
-			Namespace: job.ObjectMeta.Namespace,
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		if err != nil && apierrs.IsNotFound(err) {
-			return
-		}
-		Expect(err).ToNot(HaveOccurred())
-
-		// Remove finalizer so that the job can be deleted.
-		job.Finalizers = []string{}
-		err = k8sClient.Update(context.Background(), job)
-		Expect(err).ToNot(HaveOccurred())
+	JustBeforeEach(func() {
+		request := CreateReconciliationRequest("non-existent-name", "namespace")
 
-		// Assert deleted.
-		err = k8sClient.Get(context.Background(), types.NamespacedName{
-			Namespace: job.ObjectMeta.Namespace,
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		Expect(err).To(HaveOccurred())
-		Expect(apierrs.IsNotFound(err)).To(Equal(true))
+		reconcileResult, reconcileError = reconciler.Reconcile(request)
 	})
 
-	// Sagemaker returns 400 for throttlingException
-	It("requeue after interval if error is 429", func() {
-		// Setup mock responses.
-		sageMakerClient := builder.
-			AddDescribeHyperParameterTuningJobErrorResponseWithMessage("ThrottlingException", 400, "request id", "Rate exceeded").
-			Build()
-
-		// Instantiate controller and reconciliation request.
-		controller := createHpoReconcilerForSageMakerClient(k8sClient, sageMakerClient, "1s")
-		request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
-
-		// Run test and verify expectations.
-		reconciliationResult, err := controller.Reconcile(request)
+	Context("No error with the K8s client", func() {
+		BeforeEach(func() {
+			reconciler = createReconciler(k8sClient, sageMakerClient, "1s", mockTrackingHPOTrainingJobSpawner{})
+		})
 
-		Expect(err).ToNot(HaveOccurred())
-		// Expect requeue after interval
-		Expect(reconciliationResult.Requeue).To(Equal(false))
-		Expect(reconciliationResult.RequeueAfter).To(Equal(controller.PollInterval))
-		Expect(receivedRequests.Len()).To(Equal(1))
+		It("should not requeue", func() {
+			ExpectNoRequeue(reconcileResult, reconcileError)
+		})
 	})
 
-	It("should update status and not requeue if error is 4xx", func() {
+	Context("An error occurred with the K8s client", func() {
+		BeforeEach(func() {
+			mockK8sClient := FailToGetK8sClient{}
+			reconciler = createReconciler(mockK8sClient, sageMakerClient, "1s", mockTrackingHPOTrainingJobSpawner{})
```

Comment:
nit: can deduplicate this code by providing an overwritable k8sClient for the `BeforeEach` functions, and move this line to `JustBeforeEach`. See the other tests (line 118 in this file) for how I did it.